### PR TITLE
Add cookiecutter template

### PR DESCRIPTION
**Pull request by bkreynen**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# FAIR Python script template (QSMM level 2)
+
+This repository provides a cookiecutter template to get started with a FAIR python script. This corrseponds to a QSSM level 2 project within QuTech.
```
Comment (QFer) [inline README.md]: QSSM->QSMM

File: README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+cookiecutter gh:QuTech-Delft/fair-python-script-template
+```
+
+You will get asked a few questions about the project you want to create. After answering these questions the template will be created in the current directory.
+
+## Developing with the template
+
+You have just created an instance of the template for your project. What now?
+
+First setup a repository for your project. Discuss with your supervisor what the best place for this is (some research groups have their own groups for this on either gitlab or github). Once you have created this repository you can push the code you just created to this repository. To do this run the following commands:
```
Comment (QFer) [inline README.md]: setup -> set up

File: README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+```bash
+git init
+git add .
+git commit -m "Initial commit"
+git remote add origin <url of your repository>
+git push -u origin master
```
Comment (QFer) [inline README.md]: Some people have chosen to (re)name master to main. Actually here we want the default branch name. Add a note how to lookup the default branch in the repository.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+```
+
+You can get started developing your own code! Start by opening the README.md file of your new repository. You will notice that additionally some of the values you answered have been filled in in this readme. Next to that there are also instruction on how to install, use, extend and test the example script that is provided in this repository. You can use these instructions to start developing your own code. Make sure to keep updating the readme with relevant documentation for your own code and to keep pushing your changes to your remote git repositry.
```
Comment (QFer) [inline README.md]: also instruction->also instructions

Comment (QFer) [inline README.md]: (last word of paragraph) repositry->repository

File: {{cookiecutter.__project_slug}}/README.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+Extend this section with specific instructions for users of your software (e.g. what format should the input data be in? What kind of settings can be changed? ...)
+-->
+
+### Pre-requisites
+
+To use this script you first need to install [python](https://www.python.org/) (version 3.10 or higher) and [poetry](https://python-poetry.org/). The installation guide also assumes you have [git](https://git-scm.com/) installed.
+
+### Installation
+
+To install this software first make a local copy of this repostitory (e.g by cloning it with `git clone`). In the root directory of your local folder run the following commands:
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: repostitory->repository

Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: Maybe make the second sentence a separate paragraph. This poetry install command must be done after cloning the repo succesfully.

File: {{cookiecutter.__project_slug}}/README.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+``` bash
+poetry install
+```
+
+### Using the script
+
+#### The input
+
+The script expects a csv file as input. The first column of this file should be the names of the columns.
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: The first row of this file...

File: {{cookiecutter.__project_slug}}/README.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+#### Settings
+
+The script has a few settings which can be changed. These settings can be found at the top of the file `{{ cookiecutter.module_name }}/main.py`.
+
+| Setting | Explanation | Default |
+| --- |  --- | --- |
+| INPUT_FILE_PATH | The script looks for the input file in this location. | './example_input.csv' |
+| OUTPUT_PATH | The script creates the output file in this location. | './output.csv' |
+| COLUMN_NAME | The name of the column on the filtering should happen. | 'column_1' |
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: ...column on which... ?

File: {{cookiecutter.__project_slug}}/README.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+To now run the script run the following command:
+
+```bash
+python -m {{ cookiecutter.module_name }}.main
+```
+
+### The output
+
+The output is a new csv file with the rows filtered out. g.The file can be found in the location defined by the `OUTPUT_PATH` setting.
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: . .g. ? something missing here?

File: {{cookiecutter.__project_slug}}/README.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+## Developer Documentation
+<!--
+Provides some information for people who want to continue developing this software (or for yourself in the future if you ever return to it). It should provide the commnands a developer needs to for example test the software. Extend this section with specific instructions for developers of your software.
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: commnands -> commands

File: {{cookiecutter.__project_slug}}/README.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+To check if new code adheres to our code style you can run flake8 with the following command:
+
+```bash
+flake8 .
+```
+
+### Managing dependencies
+
+To add a new package that is reqequired use:
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: spelling: required

File: {{cookiecutter.__project_slug}}/README.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+poetry add <package_name>
+```
+
+To update the dependencies run:
+
+```bash
+poetry update
+```
+
+If you want other people to be able to run the script with the same updated packages (for exampel to reproduce your results) then make sure to commit the changes to your `poetry.lock` file.
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: exampel->example

Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: I can imagine people do not know exactly what poetry lock file is and how to commit the changes to it

File: {{cookiecutter.__project_slug}}/README.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+## Software Development Plan
+
+This sections describes how we will develop the software. It covers aspects like documentation, maintenance and dependency management.
+
+### Dependencies
+<!--
+Describe the depedencies your project needs. Before using dependencies consider whether they are of a good enough quality for your research and whether their license allows them to be used by your.
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: by your? "by your code" or "by you"

File: {{cookiecutter.__project_slug}}/README.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+* Pandas: for easier data processing. Pandas is widely used, actively maintained and has a 3-clause BSD license allowing us to use and modify it.
+
+#### Dependency Management
+<!--
+Describe how you want to manage depedencies. There are various tools available. What is important is that for the packages you import you define:
+1. The names of the packages
+2. The versions of the packages you have used
+That way someone new (or you yourself in the future) can easily install those same packages with the correct versions. In Python tools like poetry or pip can help with this.
+-->
+We use Poetry to manage the depdencies for us. The poetry.lock file is committed so that it is clear which specific versions were used in a certain version of this script.
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: spelling: depdencies

File: {{cookiecutter.__project_slug}}/README.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+### Testing
+<!--
+What and how do you test it? E.g. do you have automatic unit tests, code coverage checks, tests for code style etc.
+-->
+We provide some simple tests to check that the processing works as expected. Additionally we use flake8 to make sure we adhere to pep8 code style standards this ensures the code remains readable. Because this is only worked on by a single developer these checks are simply ran manually. The commands to run it can be found in the developer documentation.
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: Because this is -> Because this code is

File: {{cookiecutter.__project_slug}}/README.md [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+
+### Version Control and Versioning
+<!--
+Describe how you will use version control and versioning. For simple scripts it can remain simple. If you want to do more complex things (e.g. proper semantic versioning for a package that you want to publish externally you can ask the SDST for advice.)
+
+-->
+We will use git as a version control system. Becasue the project is being worked on by a single developer we keep the flow simple. Once changes are ready they are committed to the main branch. If we want to refer to a specific version of the software (e.g. a specific version used in a certain analysis) then we refer to the specific commit that was used.
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: Spelling: Becasue

File: {{cookiecutter.__project_slug}}/README.md [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+### Safety, Securiy, Privacy Assurance
+<!--
+If it could be an issue describe safety, security and privacy considerations.
+-->
+This software does not process any personal data and does not send any data to 3rd parties we therefore do not foresee safety, security or privacy issues.
+
+### Software licensing
+<!--
+What licensing will your software use? For private projects this is not very relevant. Without a license they will be considere property of QuTech. For public repositories this is very important to consider. If you make your software public the license will determine who and how the software can be used. This also has implications for intelectual property (copyright).
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/README.md]: considere -> considered

Comment (QFer) [reply {{cookiecutter.__project_slug}}/README.md]: intelectual->intellectual

File: {{cookiecutter.__project_slug}}/{{cookiecutter.module_name}}/main.py [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+OUTPUT_PATH = './output.csv'
+THRESHOLD = 5
+COLUMN_NAME = 'column_1'
+
+try:
+    data = pd.read_csv(INPUT_FILE_PATH)
+except FileNotFoundError:
+    print(f"File '{INPUT_FILE_PATH}' not found. Make sure the file exists.")
+    exit()
+
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/{{cookiecutter.module_name}}/main.py]: consider using try...except....else here instead if calling exit()

File: {{cookiecutter.__project_slug}}/{{cookiecutter.module_name}}/processing.py [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def filter_outliers(data: DataFrame,
+                    column_to_filter: str,
+                    threshold: int) -> DataFrame:
+    '''
+    Filters out all rows from data which have a value larger
+    than the threshold.
+      Parameters:
+        data (DataFrame): Dataframe to filter
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/{{cookiecutter.module_name}}/processing.py]: the types (paratheses) of the parameters are already clear from the method parameters.

**Review by QFer — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (bkreynen) [reply README.md]: Good point, will do.

File: {{cookiecutter.__project_slug}}/{{cookiecutter.module_name}}/main.py (hunk 17 above)
Comment (bkreynen) [reply {{cookiecutter.__project_slug}}/{{cookiecutter.module_name}}/main.py]: Great tip, changed it.

File: {{cookiecutter.__project_slug}}/{{cookiecutter.module_name}}/processing.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
+from pandas import DataFrame
+
+
+def filter_outliers(data: DataFrame,
+                    column_to_filter: str,
+                    threshold: int) -> DataFrame:
+    '''
```
Comment (QFer) [inline {{cookiecutter.__project_slug}}/{{cookiecutter.module_name}}/processing.py:7]: Shouldn't these be double quotes

**Review by QFer — APPROVED** [review]

**Pull request merged**